test/TestUserAcceptance.hs

```diff
@@ -42,198 +49,138 @@ testTimeout = 10 ^ 6 * 8
-              False
-              "body"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Escape"
-              "exit insert mode in vim"
-              False
-              "body"
-              assertSubstrInOutput
-        , ApplicationStep
-              ": x\r"
-              "exit vim"
-              False
-              "Attachments"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Tab"
-              "switch back to index"
-              False
-              "Testmail"
-              assertSubstrInOutput
-        , ApplicationStep
-              "Tab"
-              "switch back to the compose editor"
-              False
-              "test subject"
-              assertSubstrInOutput]
-
-data ApplicationStep = ApplicationStep
-    { asKeys :: String  -- ^ the actual commands to send
-    , asDescription :: String  -- ^ step definition
-    , asAsLiteralKey :: Bool  -- ^ disables key name lookup and sends literal input
-    , asExpected :: String  -- ^ wait until the terminal shows the expected string or timeout
-    , asAssertInOutput :: String -> String -> Assertion  -- ^ assert this against the snapshot
-    }
-
-assertSubstrInOutput :: String -> String -> Assertion
-assertSubstrInOutput out substr = assertBool "in out" $ substr `isInfixOf` out
+  withTmuxSession "user can switch back to mail index during composition" $
+        \step -> do
+            liftIO $ step "start composition"
+            out <- sendKeys "m" "From"
+            assertSubstrInOutput "From" out
+
+            liftIO $ step "enter from email"
+            out <- sendKeys "testuser@foo.test\r" "To"
+            assertSubstrInOutput "To" out
+
+            liftIO $ step "enter to: email"
+            out <- sendKeys  "user@to.test\r" "Subject"
+            assertSubstrInOutput "Subject" out
+
+            liftIO $ step "enter subject"
+            out <- sendKeys  "test subject\r" "~"
+            assertSubstrInOutput "~" out
+
+            liftIO $ step "enter mail body"
+            out <- sendKeys  "iThis is a test body" "body"
+            assertSubstrInOutput "body" out
+
+            liftIO $ step "exit insert mode in vim"
+            out <- sendKeys  "Escape" "body"
+            assertSubstrInOutput "body" out
+
+            liftIO $ step "exit vim"
+            out <- sendKeys  ": x\r" "Attachments"
+            assertSubstrInOutput "Attachments" out
+
+            liftIO $ step "switch back to index"
+            out <- sendKeys  "Tab" "Testmail"
+            assertSubstrInOutput "Testmail" out
+
+            liftIO $ step "switch back to the compose editor"
+            out <- sendKeys  "Tab" "test subject"
+            assertSubstrInOutput "test subject" out
+
+type Env = (String, String, String)
+
+assertSubstrInOutput :: String -> String -> ReaderT Env IO ()
+assertSubstrInOutput substr out = liftIO $ assertBool (substr ++ " not found in\n\n" ++ out) $ substr `isInfixOf` out
```

Comment:
If we're touching these lines anyway, can we replace applications of `(++)` with
`Data.Semigroup.<>`?

Just a nit, I don't really mind if we don't do it :)

Reply:
Good point. Will do.